src/steamship/agents/parsers/llm.py

```diff
@@ -1,37 +1,33 @@
-from typing import Tuple, List
+import re
+from typing import Dict, Optional, Union
 
 from steamship import Block
+from steamship.agents.base import Action, AgentContext, BaseTool, FinishAction
 from steamship.agents.parsers.base import OutputParser
+from steamship.experimental.transports.chat import ChatMessage
 
 
 class LLMToolOutputParser(OutputParser):
-    def remove_prefix(self, text: str, prefix: str):
-        return text[text.startswith(prefix) and len(prefix):]
-
-    # note: I do not like this construction. signature needs tweaking.
-    def parse(self, llm_generation: str) -> Tuple[str, List[Block]]:
-        blocks = []
-        lines = llm_generation.split("\n")
-        print(f"LLM Generation: {lines}")
-        tool = None
-        for line in lines:
-            if "do i need to use a tool?" in line.lower() and " no" in line.lower():
-                print("found finish!")
-                tool = "__finish__"
-                continue
-
-            if "ai: " in line.lower():
-                blocks.append(Block(text=self.remove_prefix(line, "AI:").strip()))
-                continue
-
-            if "action:" in line.lower():
-                tool = self.remove_prefix(line, "Action:").strip()
-                continue
-
-            if "action input:" in line.lower():
-                # todo: convert Block(UUID) -> Block
-                action_input = self.remove_prefix(line, "Action Input:").strip()
-                blocks.append(Block(text=action_input))
-
-        print(f"Action tool: {tool or 'no tool'}")
-        return tool, blocks
+    tools_lookup_dict: Optional[Dict[str, BaseTool]] = None
+
+    def __init__(self, **kwargs):
+        tools_lookup_dict = {tool.name: tool for tool in kwargs.pop("tools", None)}
+        super().__init__(tools_lookup_dict=tools_lookup_dict, **kwargs)
+
+    def parse(self, text: str, context: AgentContext) -> Union[Action, FinishAction]:
```

Comment:
Not a blocking comment:

I've been wondering about `Union[Action, FinishAction]`

For future discussion -- do folks se the world as only having two types of actions from the python class perspective, or a ton of base classes and `Action` just stands in as the one base-class to represent them all?